include/lock/optimistic_lock.hpp

```diff
@@ -151,6 +151,24 @@ class OptimisticLock
     }
   }
 
+  /**
+   * @brief Get an exclusive lock if version is same.
+   *
+   * @param ver an expected version value (turned off by kSIXAndSBitsMask).
+   * @retval true if the given version value is the same as a current one.
+   * @retval false otherwise.
+   */
+  [[nodiscard]] auto
+  TryLockX(const uint64_t ver)  //
+      -> bool
+  {
+    auto expected = ver;
+    const auto desired = (lock_.load(std::memory_order_relaxed) & kAllBitsMask) | kXLock;
+    const auto cas_success = lock_.compare_exchange_weak(
+        expected, desired, std::memory_order_acquire, std::memory_order_relaxed);
+    return cas_success;
```

Comment:
他のスレッドによって排他ロックが取得されたりバージョンが上がったりしない限りはリトライしてOKだと思います．

```suggestion
    auto expected = ver;
    const auto desired = ver | kXLock;
    while (true) {
      for (size_t i = 1; true; ++i) {
        const auto cas_success = lock_.compare_exchange_weak(
            expected, desired, std::memory_order_acquire, std::memory_order_relaxed);
        if (cas_success) return true;
        if ((expected & kSIXAndSBitsMask) != ver) return false;
        if (i >= kRetryNum) break;

        expected = ver;
        CPP_UTILITY_SPINLOCK_HINT
      }

      std::this_thread::sleep_for(kShortSleep);
    }
```

たぶん同じ感じで`TryLockSIX`も実装できると思うので，そちらも組んでみてもらえますか？OSLで使うタイミングがありそうな気がしています．

Reply:
たぶん，`TryLockSIX`は

```cpp
expected = ver | (lock_.load(std::memory_order_relaxed) & ~kSBitsMask);
desired = expected | kSIXLock;
```

になる感じですかね．